File: go/axonium/catalog.go
package axonium

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"sync"
)

// Model is one entry of the gateway's catalog.
type Model struct {
	ID       string   `json:"id"`
	Object   string   `json:"object"`
	OwnedBy  string   `json:"owned_by"`
	Modality string   `json:"modality"`
	Aliases  []string `json:"aliases,omitempty"`

	Raw map[string]any `json:"-"`
}

// ModelList is a catalog response.
type ModelList struct {
	Object string  `json:"object"`
	Data   []Model `json:"data"`

	Meta ResponseMeta `json:"-"`

	// Raw is the decoded body as received, so backend-specific fields this SDK does
	// not model stay reachable rather than being dropped.
	Raw map[string]any `json:"-"`
}

func (l *ModelList) setRaw(raw map[string]any) { l.Raw = raw }

// IDs returns just the model identifiers.
func (l *ModelList) IDs() []string {
	ids := make([]string, 0, len(l.Data))
	for _, m := range l.Data {
		ids = append(ids, m.ID)
	}
	return ids
}

// Find returns the model with the given ID, matching aliases too.
func (l *ModelList) Find(id string) (*Model, bool) {
	for i := range l.Data {
		if l.Data[i].ID == id {
			return &l.Data[i], true
		}
		for _, alias := range l.Data[i].Aliases {
			if alias == id {
				return &l.Data[i], true
			}
		}
	}
	return nil, false
}

// ModelsService is the model catalog API.
type ModelsService struct {
	client *Client

	mu     sync.Mutex
	mine   *ModelList
	cached bool
}

// List returns the full public catalog. This endpoint needs no authentication, so it works before
// any credential is configured -- useful for checking connectivity.
func (s *ModelsService) List(ctx context.Context) (*ModelList, error) {
	var out ModelList
	meta, err := s.client.doJSON(ctx, http.MethodGet, "/v1/models", nil, &out)
	if err != nil {
		return nil, err
	}
	out.Meta = meta
	return &out, nil
}

// Mine returns the models this token is actually allowed to call.
//
// Access is deny-by-default and granted per model, so the public catalog is not the answer to
// "what can I call": this endpoint is. The result is cached for the client's lifetime, since scope
// grants do not change within one process's run.
func (s *ModelsService) Mine(ctx context.Context) (*ModelList, error) {
	s.mu.Lock()
	if s.cached {
		defer s.mu.Unlock()
		return s.mine, nil
	}
	s.mu.Unlock()

	var out ModelList
	meta, err := s.client.doJSON(ctx, http.MethodGet, "/v1/models/mine", nil, &out)
	if err != nil {
		return nil, err
	}
	out.Meta = meta

	s.mu.Lock()
	s.mine, s.cached = &out, true
	s.mu.Unlock()
	return &out, nil
}

// Modality sets accepted by each endpoint.
var (
	modalitiesChat       = []string{"text", "vision"}
	modalitiesEmbeddings = []string{"embedding"}
	modalitiesImages     = []string{"image"}
)

// checkModality verifies a model's modality against the endpoint before sending.
//
// The gateway's own check is one-directional: calling /v1/embeddings with a text model is
// rejected, but calling /v1/chat/completions with an embedding model is not -- it returns 200 with
// degenerate output that is billed. This closes that gap, and catches typos, before the request is
// sent.
//
// Off unless Config.VerifyModality is set, because it costs one catalog request per client and the
// SDK otherwise makes no request a caller did not ask for. If the catalog cannot be loaded the
// check is skipped rather than failing the request: a guard rail must not become a new way for
// inference to break.
func (c *Client) checkModality(ctx context.Context, model string, accepted []string) error {
	if !c.config.VerifyModality {
		return nil
	}

	catalog, err := c.Models.List(ctx)
	if err != nil {
		return nil
	}
	found, ok := catalog.Find(model)
	if !ok {
		return fmt.Errorf("%w: model %q is not in the gateway's catalog; known models: %s",
			ErrInvalidRequest, model, strings.Join(catalog.IDs(), ", "))
	}
	if found.Modality == "" {
		return nil
	}
	for _, a := range accepted {
		if found.Modality == a {
			return nil
		}
	}
	return fmt.Errorf("%w: model %q has modality %q, which this endpoint does not accept (it takes %s); the gateway would answer 200 with degenerate output and bill for it",
		ErrInvalidRequest, model, found.Modality, strings.Join(accepted, " or "))
}

File: rust/src/catalog.rs
//! The model catalog, and the optional modality preflight.

use serde::Deserialize;
use serde_json::Value;

use crate::client::{CallOptions, Client};
use crate::error::{Error, Result};
use crate::types::ResponseMeta;

/// Modalities each endpoint can serve. `text` and `vision` are BOTH served on the chat endpoint,
/// so the mapping to a consumer's own vocabulary is not one to one.
pub(crate) const CHAT_MODALITIES: &[&str] = &["text", "vision"];
pub(crate) const EMBEDDING_MODALITIES: &[&str] = &["embedding"];
pub(crate) const IMAGE_MODALITIES: &[&str] = &["image"];

/// Every modality this build can place. Anything outside it is newer than this build and is not
/// ours to reject -- the platform added two in a single week, and a guard rail that starts
/// refusing valid requests each time would be worse than no guard rail.
const KNOWN_MODALITIES: &[&str] = &["text", "vision", "embedding", "image"];

/// One entry of the catalog.
#[derive(Debug, Clone, Deserialize)]
pub struct Model {
    pub id: String,
    #[serde(default)]
    pub modality: String,
    /// `None` for image models, which have no context window at all. A zero would read as a window
    /// of zero and make a `prompt_tokens < context_length` check reject every image request.
    ///
    /// For a model served by several instances this is the **smallest** of them, so a request that
    /// fits the advertised number fits whichever instance serves it.
    #[serde(default)]
    pub context_length: Option<u64>,
    /// How many instances currently serve this model. Informational: an instance is never
    /// addressed through `model`.
    #[serde(default)]
    pub served_by: Option<u32>,
    #[serde(default)]
    pub family: String,
    #[serde(default)]
    pub quantization: String,
    #[serde(default)]
    pub owned_by: String,
}

/// A catalog response.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct ModelList {
    #[serde(default)]
    pub data: Vec<Model>,
    #[serde(skip)]
    pub raw: Value,
    #[serde(skip)]
    pub meta: ResponseMeta,
}

impl ModelList {
    pub fn ids(&self) -> Vec<&str> {
        self.data.iter().map(|m| m.id.as_str()).collect()
    }

    pub fn find(&self, id: &str) -> Option<&Model> {
        self.data.iter().find(|m| m.id == id)
    }
}

impl Client {
    /// The full public catalog. Needs no authentication, so it works before any credential is
    /// configured -- useful for checking connectivity.
    pub async fn models(&self) -> Result<ModelList> {
        self.catalog_request("/v1/models").await
    }

    /// The models this token is actually allowed to call.
    ///
    /// Access is deny-by-default and granted per model, so the public catalog does not answer
    /// "what can I call": this does.
    pub async fn models_mine(&self) -> Result<ModelList> {
        self.catalog_request("/v1/models/mine").await
    }

    async fn catalog_request(&self, path: &str) -> Result<ModelList> {
        let (response, meta) = self.send(path, None, &CallOptions::default()).await?;
        let raw: Value = response.json().await.map_err(|e| {
            Error::Transport(format!(
                "the gateway returned a catalog this SDK could not parse: {e}"
            ))
        })?;
        let mut list: ModelList = serde_json::from_value(raw.clone()).map_err(|e| {
            Error::Transport(format!(
                "the gateway returned a catalog this SDK could not parse: {e}"
            ))
        })?;
        list.raw = raw;
        list.meta = meta;
        Ok(list)
    }

    /// Verifies a model's modality against the endpoint before sending.
    ///
    /// The gateway used to accept chat on an embedding model and answer `200` with degenerate,
    /// billable output; it now rejects it, so this is a typo-catcher and a saved round trip rather
    /// than a correctness guard. Off unless `verify_modality` is set, because it costs one catalog
    /// request and the SDK otherwise makes no request a caller did not ask for.
    ///
    /// If the catalog cannot be loaded the check is skipped: a guard rail must not become a new
    /// way for inference to fail.
    pub(crate) async fn check_modality(&self, model: &str, accepted: &[&str]) -> Result<()> {
        if !self.config.verify_modality {
            return Ok(());
        }
        let Ok(catalog) = self
            .catalog
            .get_or_try_init(|| async { self.models().await })
            .await
        else {
            return Ok(());
        };

        let Some(found) = catalog.find(model) else {
            return Err(Error::InvalidRequest(format!(
                "model {model:?} is not in the gateway's catalog; known models: {}",
                catalog.ids().join(", ")
            )));
        };
        if found.modality.is_empty() || !KNOWN_MODALITIES.contains(&found.modality.as_str()) {
            return Ok(());
        }
        if accepted.contains(&found.modality.as_str()) {
            return Ok(());
        }
        Err(Error::InvalidRequest(format!(
            "model {model:?} has modality {:?}, which this endpoint does not accept (it takes {})",
            found.modality,
            accepted.join(" or ")
        )))
    }
}
